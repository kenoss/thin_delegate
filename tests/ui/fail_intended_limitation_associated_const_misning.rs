// Compare with pass_fill_missing_functions_in_impl.rs
//
// You need to fill associated consts. thin_delegate doesn't automatically fill them because there
// is no natural choise in general, e.g. enum case.

#[thin_delegate::register]
trait Hello {
    type Return;

    const HAS_DEFAULT: &'static str = "HAS_DEFAULT";
    const NEED_TO_FILL: &'static str;

    // `thin_delegate` only can fill associated functions.
    fn filled(&self) -> Self::Return;
    fn override_(&self) -> Self::Return;
}

impl Hello for String {
    type Return = String;

    const NEED_TO_FILL: &'static str = "String";

    fn filled(&self) -> Self::Return {
        self.clone()
    }

    fn override_(&self) -> Self::Return {
        self.clone()
    }
}

#[thin_delegate::register]
struct Hoge(String);

#[thin_delegate::fill_delegate]
impl Hello for Hoge {
    type Return = String;

    // const NEED_TO_FILL: &'static str = "Hoge";

    fn override_(&self) -> Self::Return {
        self.0.override_().to_uppercase()
    }
}

fn main() {}
