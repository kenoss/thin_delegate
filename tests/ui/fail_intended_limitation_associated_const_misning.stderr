error[E0046]: not all trait items implemented, missing: `NEED_TO_FILL`
  --> tests/ui/fail_intended_limitation_associated_const_misning.rs:36:1
   |
11 |     const NEED_TO_FILL: &'static str;
   |     -------------------------------- `NEED_TO_FILL` from trait
...
36 | impl Hello for Hoge {
   | ^^^^^^^^^^^^^^^^^^^ missing `NEED_TO_FILL` in implementation
